File: API/2154_MACD_Trend_Mode/PY/macd_trend_mode_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import MovingAverageConvergenceDivergenceSignal
from StockSharp.Algo.Strategies import Strategy


class macd_trend_mode_strategy(Strategy):
    # TrendModes: 0=Histogram, 1=Cloud, 2=Zero
    def __init__(self):
        super(macd_trend_mode_strategy, self).__init__()
        self._fast_length = self.Param("FastLength", 12) \
            .SetDisplay("Fast Length", "Fast EMA period", "MACD")
        self._slow_length = self.Param("SlowLength", 26) \
            .SetDisplay("Slow Length", "Slow EMA period", "MACD")
        self._signal_length = self.Param("SignalLength", 9) \
            .SetDisplay("Signal Length", "Signal line period", "MACD")
        self._trend_mode = self.Param("TrendMode", 1) \
            .SetDisplay("Trend Mode", "0=Histogram, 1=Cloud, 2=Zero", "General")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._prev_hist = 0.0
        self._prev_prev_hist = 0.0
        self._has_prev_hist = False
        self._has_prev_prev_hist = False
        self._prev_macd = 0.0
        self._prev_signal = 0.0
        self._has_prev_lines = False

    @property
    def fast_length(self):
        return self._fast_length.Value

    @property
    def slow_length(self):
        return self._slow_length.Value

    @property
    def signal_length(self):
        return self._signal_length.Value

    @property
    def trend_mode(self):
        return self._trend_mode.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(macd_trend_mode_strategy, self).OnReseted()
        self._prev_hist = 0.0
        self._prev_prev_hist = 0.0
        self._has_prev_hist = False
        self._has_prev_prev_hist = False
        self._prev_macd = 0.0
        self._prev_signal = 0.0
        self._has_prev_lines = False

    def OnStarted2(self, time):
        super(macd_trend_mode_strategy, self).OnStarted2(time)
        macd = MovingAverageConvergenceDivergenceSignal()
        macd.Macd.ShortMa.Length = self.fast_length
        macd.Macd.LongMa.Length = self.slow_length
        macd.SignalMa.Length = self.signal_length
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.BindEx(macd, self.process_candle).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, macd)
            self.DrawOwnTrades(area)

    def process_candle(self, candle, macd_value):
        if candle.State != CandleStates.Finished:
            return
        if not macd_value.IsFinal or not macd_value.IsFormed:
            return
        macd_val_raw = macd_value.Macd
        signal_val_raw = macd_value.Signal
        if macd_val_raw is None or signal_val_raw is None:
            return
        macd_val = float(macd_val_raw)
        signal_val = float(signal_val_raw)
        hist = macd_val - signal_val
        buy_signal = False
        sell_signal = False
        mode = int(self.trend_mode)
        if mode == 0:  # Histogram
            if self._has_prev_hist and self._has_prev_prev_hist:
                if self._prev_hist < self._prev_prev_hist and hist > self._prev_hist:
                    buy_signal = True
                if self._prev_hist > self._prev_prev_hist and hist < self._prev_hist:
                    sell_signal = True
            self._prev_prev_hist = self._prev_hist
            self._has_prev_prev_hist = self._has_prev_hist
            self._prev_hist = hist
            self._has_prev_hist = True
        elif mode == 1:  # Cloud
            if self._has_prev_lines:
                if self._prev_macd <= self._prev_signal and macd_val > signal_val:
                    buy_signal = True
                if self._prev_macd >= self._prev_signal and macd_val < signal_val:
                    sell_signal = True
            self._prev_macd = macd_val
            self._prev_signal = signal_val
            self._has_prev_lines = True
        elif mode == 2:  # Zero
            if self._has_prev_hist:
                if self._prev_hist <= 0 and hist > 0:
                    buy_signal = True
                if self._prev_hist >= 0 and hist < 0:
                    sell_signal = True
            self._prev_hist = hist
            self._has_prev_hist = True
        if buy_signal and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        elif sell_signal and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()

    def CreateClone(self):
        return macd_trend_mode_strategy()

File: API/2101-2200/2154_MACD_Trend_Mode/CS/MacdTrendModeStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Ecng.Common;
using Ecng.Collections;
using Ecng.Serialization;
using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// MACD strategy with selectable trend detection modes.
/// </summary>
public class MacdTrendModeStrategy : Strategy
{
	public enum TrendModes { Histogram, Cloud, Zero }

	private readonly StrategyParam<int> _fastLength;
	private readonly StrategyParam<int> _slowLength;
	private readonly StrategyParam<int> _signalLength;
	private readonly StrategyParam<TrendModes> _trendMode;
	private readonly StrategyParam<DataType> _candleType;

	private decimal _prevHist, _prevPrevHist;
	private bool _hasPrevHist, _hasPrevPrevHist;
	private decimal _prevMacd, _prevSignal;
	private bool _hasPrevLines;

	public int FastLength { get => _fastLength.Value; set => _fastLength.Value = value; }
	public int SlowLength { get => _slowLength.Value; set => _slowLength.Value = value; }
	public int SignalLength { get => _signalLength.Value; set => _signalLength.Value = value; }
	public TrendModes TrendMode { get => _trendMode.Value; set => _trendMode.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public MacdTrendModeStrategy()
	{
		_fastLength = Param(nameof(FastLength), 12).SetGreaterThanZero()
			.SetDisplay("Fast Length", "Fast EMA period", "MACD");
		_slowLength = Param(nameof(SlowLength), 26).SetGreaterThanZero()
			.SetDisplay("Slow Length", "Slow EMA period", "MACD");
		_signalLength = Param(nameof(SignalLength), 9).SetGreaterThanZero()
			.SetDisplay("Signal Length", "Signal line period", "MACD");
		_trendMode = Param(nameof(TrendMode), TrendModes.Cloud)
			.SetDisplay("Trend Mode", "Trend detection mode", "General");
		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(4).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType)];

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_prevHist = default;
		_prevPrevHist = default;
		_hasPrevHist = false;
		_hasPrevPrevHist = false;
		_prevMacd = default;
		_prevSignal = default;
		_hasPrevLines = false;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var macd = new MovingAverageConvergenceDivergenceSignal
		{
			Macd = { ShortMa = { Length = FastLength }, LongMa = { Length = SlowLength } },
			SignalMa = { Length = SignalLength }
		};

		var subscription = SubscribeCandles(CandleType);
		subscription.BindEx(macd, ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, macd);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, IIndicatorValue macdValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!macdValue.IsFinal || !macdValue.IsFormed)
			return;

		var macdTyped = (IMovingAverageConvergenceDivergenceSignalValue)macdValue;
		if (macdTyped.Macd is not decimal macdVal || macdTyped.Signal is not decimal signalVal)
			return;

		var hist = macdVal - signalVal;
		var buySignal = false;
		var sellSignal = false;

		switch (TrendMode)
		{
			case TrendModes.Histogram:
				if (_hasPrevHist && _hasPrevPrevHist)
				{
					if (_prevHist < _prevPrevHist && hist > _prevHist) buySignal = true;
					if (_prevHist > _prevPrevHist && hist < _prevHist) sellSignal = true;
				}
				_prevPrevHist = _prevHist;
				_hasPrevPrevHist = _hasPrevHist;
				_prevHist = hist;
				_hasPrevHist = true;
				break;

			case TrendModes.Cloud:
				if (_hasPrevLines)
				{
					if (_prevMacd <= _prevSignal && macdVal > signalVal) buySignal = true;
					if (_prevMacd >= _prevSignal && macdVal < signalVal) sellSignal = true;
				}
				_prevMacd = macdVal;
				_prevSignal = signalVal;
				_hasPrevLines = true;
				break;

			case TrendModes.Zero:
				if (_hasPrevHist)
				{
					if (_prevHist <= 0m && hist > 0m) buySignal = true;
					if (_prevHist >= 0m && hist < 0m) sellSignal = true;
				}
				_prevHist = hist;
				_hasPrevHist = true;
				break;
		}

		if (buySignal && Position <= 0)
		{
			if (Position < 0) BuyMarket();
			BuyMarket();
		}
		else if (sellSignal && Position >= 0)
		{
			if (Position > 0) SellMarket();
			SellMarket();
		}
	}
}
